docs/high-level-overview.md

```diff
@@ -1,165 +1,179 @@
-conditions are met. Each delivery on the target chain is confirmed back to the source chain by the
-relayer. This is so that she can collect the reward for delivering these messages.
+As its on-chain component (which requires bridge GRANDPA pallet to be deployed nearby), the parachains
+finality relay requires GRANDPA finality relay to be running in parallel. Without it, the header `B` or
+some its ancestor won't be finalized and we'll be unable to verify generated storage proof. 
 
-Users of the pallet add their messages to an "outbound lane" on the source chain. When a block is
-finalized message relayers are responsible for reading the current queue of messages and submitting
-some (or all) of them to the "inbound lane" of the target chain. Each message has a `nonce`
-associated with it, which serves as the ordering of messages. The inbound lane stores the last
-delivered nonce to prevent replaying messages. To successfully deliver the message to the inbound lane
-on target chain the relayer has to present present a storage proof which shows that the message was
-part of the outbound lane on the source chain.
+More: [Parachains Finality Relay Sequence Diagram](./parachains-finality-relay.html), [code](../relays/parachains/).
 
-During delivery of messages they are immediately dispatched on the target chain and the relayer is
-required to declare the correct `weight` to cater for all messages dispatch and pay all required
-fees of the target chain. To make sure the relayer is incentivised to do so, on the source chain:
-- the user provides a declared dispatch weight of the payload
-- the pallet calculates the expected fee on the target chain based on the declared weight
-- the pallet converts the target fee into source tokens (based on a price oracle) and reserves
-  enough tokens to cover for the delivery, dispatch, confirmation and additional relayers reward.
+### Messages Relay
 
-If the declared weight turns out to be too low on the target chain the message is delivered but
-it immediately fails to dispatch. The fee and reward is collected by the relayer upon confirmation
-of delivery.
+Messages relay is actually a two relays that are running in a single process: messages delivery relay and
+delivery confirmation relay. Even though they are more complex and have many caveats, the overall algorithm
+is the same as in other relays.
 
-Due to the fact that message lanes require delivery confirmation transactions, they also strictly
-require bi-directional header sync (i.e. you can't use message delivery with one-way header sync).
+Message delivery relay connects to the source chain and looks at the outbound lane end, waiting until new
+messages are queued there. Once they appear at the source block `B`, the relay start waiting for the block
+`B` or its ancestor appear at the target chain. Then the messages storage proof is generated and submitted
```

Comment:
again: afaik in blockchain terms, `ancestor` means `(grand(grand..))parent` - maybe `children` or `descendants`?
```suggestion
`B` or its descendant appear at the target chain. Then the messages storage proof is generated and submitted
```